### ISSUE TITLE
test_zippath fails on Windows with Python 3.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8747]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8747|
|Type|defect|
|Created|2016-08-12 00:03:00Z|
|Branch|https://github.com/twisted/twisted/tree/8747-rodrigc-zippath-py3|

On Windows, with Python 3.5, if I do:

```
python -m twisted.trial twisted.python.test.test_zippath
```

I get:

```
twisted.python.test.test_zippath
  ZipFilePathTests
    test_childrenNonexistentError ...                                    [FAIL]
    test_childrenNotDirectoryError ...                                   [FAIL]
    test_dictionaryKeyWithString ...                                       [OK]
    test_dictionaryKeys ...                                                [OK]
    test_invalidSubdir ...                                                 [OK]
    test_multipleChildSegments ...                                         [OK]
    test_newTimesAreFloats ...                                             [OK]
    test_oldTimesAreInts ...                                               [OK]
    test_parents ...                                                       [OK]
    test_segmentsFromNegative ...                                          [OK]
    test_segmentsFromPositive ...                                          [OK]
    test_validFiles ...                                                    [OK]
    test_validSubdir ...                                                   [OK]
    test_verifyObject ...                                                  [OK]
    test_walk ...                                                          [OK]
    test_zipArchiveRepr ...                                                [OK]
    test_zipPathRepr ...                                                   [OK]
    test_zipPathReprParentDirSegment ...                                   [OK]
twisted.test.test_paths
  AbstractFilePathTests
    test_childrenNonexistentError ...                                      [OK]
    test_childrenNotDirectoryError ...                                     [OK]
    test_dictionaryKeyWithString ...                                       [OK]
    test_dictionaryKeys ...                                                [OK]
    test_invalidSubdir ...                                                 [OK]
    test_multipleChildSegments ...                                         [OK]
    test_newTimesAreFloats ...                                             [OK]
    test_oldTimesAreInts ...                                               [OK]
    test_parents ...                                                       [OK]
    test_segmentsFromNegative ...                                          [OK]
    test_segmentsFromPositive ...                                          [OK]
    test_validFiles ...                                                    [OK]
    test_validSubdir ...                                                   [OK]
    test_verifyObject ...                                                  [OK]
    test_walk ...                                                          [OK]

===============================================================================
[FAIL]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\twisted\test\test_paths.py", line 217, in test_childrenNonexistentError
    self.path.child(b'not real').children)
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 346, in __exit__
    self._expectedName, reason.getTraceback()),
twisted.trial.unittest.FailTest: builtins.FileNotFoundError raised instead of UnlistableError:
 Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\twisted\python\util.py", line 906, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "c:\Users\Craig\twisted3\twisted\test\test_paths.py", line 217, in test_childrenNonexistentError
    self.path.child(b'not real').children)
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
--- <exception caught here> ---
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in <lambda>
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\python\filepath.py", line 332, in children
    subnames = self.listdir()
  File "c:\Users\Craig\twisted3\twisted\python\zippath.py", line 133, in listdir
    raise OSError(errno.ENOENT, "Non-existent zip entry listed")
builtins.FileNotFoundError: [Errno 2] Non-existent zip entry listed


twisted.python.test.test_zippath.ZipFilePathTests.test_childrenNonexistentError
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\twisted\test\test_paths.py", line 225, in test_childrenNotDirectoryError
    self.path.child(b'file1').children)
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 346, in __exit__
    self._expectedName, reason.getTraceback()),
twisted.trial.unittest.FailTest: builtins.FileNotFoundError raised instead of UnlistableError:
 Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\twisted\python\util.py", line 906, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "c:\Users\Craig\twisted3\twisted\test\test_paths.py", line 225, in test_childrenNotDirectoryError
    self.path.child(b'file1').children)
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
--- <exception caught here> ---
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 308, in _handle
    self._returnValue = obj()
  File "c:\Users\Craig\twisted3\twisted\trial\_synctest.py", line 414, in <lambda>
    return context._handle(lambda: f(*args, **kwargs))
  File "c:\Users\Craig\twisted3\twisted\python\filepath.py", line 332, in children
    subnames = self.listdir()
  File "c:\Users\Craig\twisted3\twisted\python\zippath.py", line 133, in listdir
    raise OSError(errno.ENOENT, "Non-existent zip entry listed")
builtins.FileNotFoundError: [Errno 2] Non-existent zip entry listed


twisted.python.test.test_zippath.ZipFilePathTests.test_childrenNotDirectoryError
-------------------------------------------------------------------------------
Ran 33 tests in 1.828s

FAILED (failures=2, successes=31)
```

<details><summary>Searchable metadata</summary>

```
trac-id__8747 8747
type__defect defect
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8747_rodrigc_zippath_py3 8747-rodrigc-zippath-py3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1470960180277989 1470960180277989
changetime__1471142772094123 1471142772094123
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/pull/488

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_2"></a>|@markrwilliams commented|
|-|-|

So `OSError(errno.ENOENT, "...")` ends up raising a non-`OSError` exception on Windows.  Crazy.

Also, for posterity, `twisted.python.filepath.UnlistableError` has quite an interesting API.

In any event this patch makes complete sense.  Please merge it.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 2472c6fc99386aacb6eb430aab52a38f67c69b61
```
#!CommitTicketReference repository="" revision="2472c6fc99386aacb6eb430aab52a38f67c69b61"
Merge pull request #488 from twisted/8747-rodrigc-zippath-py3

Author: hawkowl, rodrigc
Reviewer: markrwilliams
Fixes: #8747

Fix test_zippath on Windows Python 3
```
